lib/pure/collections/queues.nim

```diff
@@ -8,56 +8,168 @@
+##     var q = initQueue[int]()  # initializes the object
+##     for i in 1 ..< a: q.add i  # populates the queue
+##
+##     if b < q.len:  # checking before indexed access
+##       echo "The element at index position ", b, " is ", q[b]
+##
+##     # The following two lines don't need any checking on access due to the
+##     # logic of the program, but that would not be the case if `a` could be 0.
+##     assert q.front == 1
+##     assert q.back == a
+##
+##     while q.len > 0:  # checking if the queue is empty
+##       echo q.pop()
+##
+##
 ## Note: For inter thread communication use
 ## a `Channel <channels.html>`_ instead.
 
+proc englishOrdinal(n: SomeInteger): string =
+  # Temporary proc. Needs to be moved somewhere else as it can be reused in
+  # other places too.
+  # If this accepted number strings instead and only gave out the letters it
+  # would be more flexible, permitting things like 1.100.000th, 34,545,321st
+  # but it would be harder and more error prone to use.
+  let num = $n
+  if num.len > 1 and num[^2] == '1':
+    return num & "th"
+  else:
+    case num[^1]
+    of '1': return num & "st"
+    of '2': return num & "nd"
+    of '3': return num & "rd"
+    else: return num & "th"
+
 import math
 
 type
-  Queue*[T] = object ## a queue
+  Queue*[T] = object ## A queue.
     data: seq[T]
     rd, wr, count, mask: int
 
 {.deprecated: [TQueue: Queue].}
 
-proc initQueue*[T](initialSize=4): Queue[T] =
-  ## creates a new queue. `initialSize` needs to be a power of 2.
+proc initQueue*[T](initialSize: int = 4): Queue[T] =
+  ## Create a new queue.
+  ## Optionally, the initial capacity can be reserved via `initialSize` as a
+  ## performance optimization. `initialSize` needs to be a power of 2.
+  ## The lenght of a newly created queue will still be 0.
   assert isPowerOfTwo(initialSize)
   result.mask = initialSize-1
   newSeq(result.data, initialSize)
 
-proc len*[T](q: Queue[T]): int =
-  ## returns the number of elements of `q`.
+proc len*[T](q: Queue[T]): int {.inline.}=
+  ## Return the number of elements of `q`.
   result = q.count
 
+proc low*[T](q: Queue[T]): int {.inline.}=
+  ## Return the index of the oldest element of `q` (always 0).
+  result = 0
+
+proc high*[T](q: Queue[T]): int {.inline.}=
+  ## Return the index of the last element inserted on `q` (equivalent to
+  ## `q.len - 1`).
+  result = q.count - 1
+
+template emptyCheck(q) =
+  # Bounds check for the regular queue access.
+  when compileOption("boundChecks"):
+    if unlikely(q.count < 1):
+      raise newException(IndexError, "Empty queue.")
+  discard
+
+template xBoundsCheck(q, i) =
+  # Bounds check for the array like accesses.
+  when compileOption("boundChecks"):  # d:release should disable this.
+    if unlikely(i >= q.count):  # x < q.low is taken care by the Natural parameter
+      raise newException(IndexError,
+                         "Tried to access the " & englishOrdinal(i+1) &
```

Comment:
I agree that it is verbose, but indexing is tricky. For example, you meant `"Out of bounds: " & $i & " > " & $(q.count-1)`. I tried to write a very clear message that nobody would come on irc asking what exactly it meant. But yeah, your version is sufficiently clear after thinking for a bit, especially if it is adopted more widely through the language.